### DISCUSSION TITLE
extremely slow/deadlocking while attempting to use pooling via `deadpool` in this crate, am I doing something wrong?

### DISCUSSION THREAD

**Discussion by vidhanio** [weiznich/diesel_async#133] [Q&A] [answered]
for context, I am working with an axum server and just running my test suite, which involves sending hundreds of requests which require the database. I am following the example specified in axum's repository (but switching bb8 for deadpool) and i have found the pool to deadlock in some cases, be flaky, and when it does work, be much slower than my previous sqlx and sea-orm rewrites of this program. is there something i'm fundamentally doing wrong?

## pool access/setup

```rust
pub struct DatabaseConnection(pub Object<AsyncPgConnection>);

#[async_trait]
impl<S> FromRequestParts<S> for DatabaseConnection
where
    S: Send + Sync,
    Pool<AsyncPgConnection>: FromRef<S>,
{
    type Rejection = ApiError;

    #[tracing::instrument(skip(state), ret, err(Debug))]
    async fn from_request_parts(
        parts: &mut ::axum::http::request::Parts,
        state: &S,
    ) -> ::std::result::Result<Self, Self::Rejection> {
        let pool = Pool::from_ref(state);

        let db = pool.get().await?;
        Ok(Self(db))
    }
}
```

```rust
        let manager = AsyncDieselConnectionManager::new(config.database_url.to_string());
        let pool = Pool::builder(manager).max_size(90).build()?;
```

## integration test setup

```rust
        let mut database_url = assert_ok!(assert_ok!(env::var("DATABASE_URL")).parse::<Url>());

        let database_name = Ulid::new().to_string();

        database_url.set_path("postgres");
        let mut db = assert_ok!(AsyncPgConnection::establish(database_url.as_str()).await);

        assert_ok!(
            sql_query(format!("CREATE DATABASE {:?}", database_name.as_str()))
                .execute(&mut db)
                .await
        );

        database_url.set_path(&database_name);

		let config = Config {
            socket: tcp_listener,
            database_url,
            jwt_secret: String::from_utf8_lossy(&rand::random::<[u8; 32]>()).into_owned(),
        };

		// api gets served (runs migrations, etc.) and the test makes http requests
```

**Comment by vidhanio** [comment]
update: i moved everything to `mobc` and it worked perfectly. is there some bug with the deadpool implementation?

**Reply by weiznich** [reply]
If you feel that this is a bug with the deadpool implementation we would appreciate a minimal reproducible example with an exact description of how to reproduce your deadlock. Ideally that is some simple project that can be cloned via git and where their problem can be reproduced by a simple cargo run command.

**Reply by vidhanio** [reply]
i'll try to get an mre in over the weekend @weiznich :) i appreciate the response.

**Comment by lsunsi** [comment]
@vidhanio What difference do you see between deadpool and mobc or even bb8? Or do you see them as commodities?
I'm moving to diesel-async slowly and opted for deadpool because of it's popularity, so I got a little worried about this discussion, even without a reproducible example.

**Reply by vidhanio** [reply, marked as answer]
Hi @lsunsi, I managed to find the source of most of my problems: I was running my tests in debug mode. Running `cargo test --release` fixed my issues with flaky tests on deadpool.
